docs/source/distributed.rst

```diff
@@ -554,7 +554,47 @@ Multi-GPU collective functions
     please contact PyTorch Distributed's maintainers.
 
 
-.. _distributed-launch:
+.. _object_collectives:
+
+Object collectives
+------------------
+
+.. warning::
+    Object collectives have a number of serious limitations.  Read further to determine
+    if they are safe to use for your use case.
+
+Object collectives are a set of collective-like operations that work on arbitrary
+python objects, as long as they can be pickled.  There are various collective patterns
+implemented (e.g. broadcast, all_gather, ...) but they each roughly follow this pattern:
+
+1. convert the input object into a pickle (raw bytes), then shove it into a byte tensor
+2. communicate the size of this byte tensor to peers (first collective operation)
+3. allocate appropriately sized tensor to perform the real collective
+4. communicate the object data (second collective operation)
+5. convert raw data back into python (unpickle)
+
+Object collectives often have surprising performance or memory characteristics that lead to
+long runtimes or OOMs, and thus they should be used with caution.  Here are some common issues.
+
+**Asymetric pickle/unpickle time** - Pickling objects can be slow, depending on the number, type and size of the objects.
+When the collective has a fan-in (e.g. gather_object), the recieving rank(s) must unpickle N times more objects than
+the sending rank(s) had to pickle.  This has led to one class of issues where users report nccl watchdog timeouts caused
+by one rank spending a really long time unpickling data.
+
+**Inefficient tensor communication** - Tensors should be sent via regular collective APIs, not object collective APIs.
+It is possible to send Tensors via object collective APIs, but they will be serialized and deserialized (including a
+cpu-sync and device-to-host copy in the case of non-cpu tensors), and in almost every case other than debugging or
+troubleshooting code, it would be worth the trouble to refactor the code to use non-object collectives instead.
+
+**Unexpected tensor devices** - If you still want to send tensors via object collectives, there is another aspect
+specific to cuda (and possibly other accelerators) tensors.  If you pickle a tensor that is currently on `cuda:3`, and
+then unpickle it, you will get another tensor on `cuda:3` *regardless of which process you are on, or which cuda device
+is the 'default' device for that process*.  With regular tensor collective APIs, 'output tensors' will always be on the
+same, local device, which is generally what you'd expect.
+
+**Unintended cuda-context creation** - Unpickling a tensor will implicity activate a cuda context if it is the first
+time a GPU is used by the process, which can waste significant amounts of GPU memory. This issue can be avoided by
+moving tensors to CPU before passing them as inputs to an object collective.
```

Comment:
It seems "Unintended cuda-context creation" is really related to "Unexpected tensor devices" above. So I recommend merging sections too. 
Could the sentence "This issue can be avoided by moving tensors to CPU before passing them as inputs to an object collective" sound like we encourage the users to do so (while we actually don't)?

Reply:
well, i said "worked around" not "should".  I think that the best I can do is outline all the risks, and if you still want to pass tensors in we can at least make it less bad.